### PR TITLE
GH-44526: [C++][Acero] Fix crash when thread in asof_join is not running

### PR DESCRIPTION
**Pull request by mroz45**
### What changes are included in this PR?

I add checking joinable before join process_thread. It will prevent exeptions in case when plan is invalid and asof_join never_starts.

* GitHub Issue: #44526

### CODE REVIEW

**Comment by zanmato1984** [comment]
Hi @mroz45 , thanks for fixing this!

I think we still need a reproducible C++ UT. If you have any difficulties on that, please let me know.

**Comment by mroz45** [comment]
@zanmato1984 
Do you think this test is enough?

**Comment by zanmato1984** [comment]
> @zanmato1984 Do you think this test is enough?

Thank you for the update. The test should suffice at my first glance. I'll take a deeper look soon.

File: cpp/src/arrow/acero/asof_join_node_test.cc [hunk 1]
```diff
@@ -1731,6 +1731,33 @@ TEST(AsofJoinTest, RhsEmptinessRaceEmptyBy) {
       ExecBatchFromJSON({int64(), utf8()}, R"([[1, "Z"], [2, "Z"], [3, "B"]])");
   AssertExecBatchesEqualIgnoringOrder(result.schema, {exp_batch}, result.batches);
 }
+// Reproduction of GH-44526: Provoke destruction of not started asofjoin node by providing
```
Comment (zanmato1984) [inline cpp/src/arrow/acero/asof_join_node_test.cc:1734]:
```suggestion

// Reproduction of GH-44526: Provoke destruction of not started asofjoin node by providing
```

File: cpp/src/arrow/acero/asof_join_node_test.cc [hunk 2]
```diff
@@ -1731,6 +1731,33 @@ TEST(AsofJoinTest, RhsEmptinessRaceEmptyBy) {
...
+  arrow::acero::SinkNodeOptions sink_node_options{nullptr, nullptr};
+  arrow::acero::Declaration sink =
+      arrow::acero::Declaration::Sequence({asof_join, {"sink", sink_node_options}});
+
+  EXPECT_RAISES_WITH_MESSAGE_THAT(
+      Invalid,
+      ::testing::HasSubstr(
+          "`generator` is a required SinkNode option and cannot be null"),
+      DeclarationToStatus(std::move(sink)));
+}
```
Comment (zanmato1984) [inline cpp/src/arrow/acero/asof_join_node_test.cc:1761]:
```suggestion
}

```

**Review by zanmato1984 — CHANGES_REQUESTED** [review]
Let's keep the blank lines around individual tests.

File: cpp/src/arrow/acero/asof_join_node_test.cc [hunk 3]
```diff
@@ -1731,6 +1731,33 @@ TEST(AsofJoinTest, RhsEmptinessRaceEmptyBy) {
...
+                                              {std::move(left_batch)})};
+  Declaration right{
+      "exec_batch_source",
+      ExecBatchSourceNodeOptions(schema({field("colVals", utf8()), field("on", int64())}),
+                                 {std::move(right_batch)})};
+  AsofJoinNodeOptions asof_join_opts({{{"on"}, {}}, {{"on"}, {}}}, 1);
+  Declaration asof_join{
+      "asofjoin", {std::move(left), std::move(right)}, std::move(asof_join_opts)};
 
+  arrow::acero::SinkNodeOptions sink_node_options{nullptr, nullptr};
```
Comment (zanmato1984) [inline cpp/src/arrow/acero/asof_join_node_test.cc:1752]: Could you add: 1) parameter name (e.g. `/*xxx=*/nullptr, /*xxx=*/nullptr`), 2) some comments about how this intentionally invalid `nullptr` is essential for the reproduction of the targeting bug?

**Review by zanmato1984 — COMMENTED** [review]
Some nits. Otherwise LGTM.

File: cpp/src/arrow/acero/asof_join_node_test.cc [hunk 4]
```diff
@@ -1732,5 +1732,37 @@ TEST(AsofJoinTest, RhsEmptinessRaceEmptyBy) {
...
+      ExecBatchSourceNodeOptions(schema({field("colVals", utf8()), field("on", int64())}),
+                                 {std::move(right_batch)})};
+  AsofJoinNodeOptions asof_join_opts({{{"on"}, {}}, {{"on"}, {}}}, 1);
+  Declaration asof_join{
+      "asofjoin", {std::move(left), std::move(right)}, std::move(asof_join_opts)};
+
+  // Setting invalid arguments, such as nullptr in generator or schema in SinkNodeOptions,
+  // causes the execution plan to terminate before the asofjoin node is started.
+  arrow::acero::SinkNodeOptions sink_node_options{/*generator=*/nullptr,
+                                                  /*schema*/ nullptr};
```
Comment (zanmato1984) [inline cpp/src/arrow/acero/asof_join_node_test.cc:1756]:
```suggestion
                                                  /*schema=*/nullptr};
```

File: .gitignore [hunk 5]
```diff
@@ -108,3 +108,10 @@ java/.mvn/.develocity/
 # rat
 filtered_rat.txt
 rat.txt
+
+/a
+/cpp/build_test/
+/cpp/src/arrow/util/config.h
+/cpp/src/parquet/parquet_version.h
+/pyarrow-dev/
+/python/myenv/
```
Comment (zanmato1984) [inline .gitignore:117]: Can we not include such changes unless absolutely necessary?

Comment (mroz45) [reply .gitignore:117]: My mistake i will correct it

**Review by zanmato1984 — APPROVED** [review]
+1

I'll keep this PR open for another few days to see if @westonpace @pitrou have any comments. After that I'll merge this.

File: cpp/src/arrow/acero/asof_join_node_test.cc [hunk 6]
```diff
@@ -1732,5 +1732,37 @@ TEST(AsofJoinTest, RhsEmptinessRaceEmptyBy) {
...
+  AsofJoinNodeOptions asof_join_opts({{{"on"}, {}}, {{"on"}, {}}}, 1);
+  Declaration asof_join{
+      "asofjoin", {std::move(left), std::move(right)}, std::move(asof_join_opts)};
+
+  // Setting invalid arguments, such as nullptr in generator or schema in SinkNodeOptions,
+  // causes the execution plan to terminate before the asofjoin node is started.
+  arrow::acero::SinkNodeOptions sink_node_options{/*generator=*/nullptr,
+                                                  /*schema=*/nullptr};
+  arrow::acero::Declaration sink =
+      arrow::acero::Declaration::Sequence({asof_join, {"sink", sink_node_options}});
```
Comment (kou) [inline cpp/src/arrow/acero/asof_join_node_test.cc:1758]: Can we simplify this?

```suggestion
  auto sink = Declaration::Sequence({asof_join, {"sink", sink_node_options}});
```

File: cpp/src/arrow/acero/asof_join_node_test.cc [hunk 7]
```diff
@@ -1732,5 +1732,36 @@ TEST(AsofJoinTest, RhsEmptinessRaceEmptyBy) {
   AssertExecBatchesEqualIgnoringOrder(result.schema, {exp_batch}, result.batches);
 }
 
+// Reproduction of GH-44526: Provoke destruction of not started asofjoin node by providing
+// a sink that fails on creation
+TEST(AsofJoinTest, DetroyNotStartedAsofJoinNode) {
```
Comment (pitrou) [inline cpp/src/arrow/acero/asof_join_node_test.cc:1737]:
```suggestion
TEST(AsofJoinTest, DestroyNonStartedAsofJoinNode) {
```

**Review by pitrou — APPROVED** [review]
Just a naming fix, otherwise LGTM

**Comment by zanmato1984** [comment]
Merging now.

**Comment by zanmato1984** [comment]
Merged. Thank you @mroz45 for fixing this!

**Comment by kou** [comment]
@zanmato1984 Could you use our merge script https://github.com/apache/arrow/tree/main/dev#how-to-merge-a-pull-request next time? It removes comments (`<!-- ... -->`) from PR description and use the prepared PR description as a commit message.

**Comment by zanmato1984** [comment]
> @zanmato1984 Could you use our merge script [`main`/dev#how-to-merge-a-pull-request](https://github.com/apache/arrow/tree/main/dev#how-to-merge-a-pull-request) next time? It removes comments (`<!-- ... -->`) from PR description and use the prepared PR description as a commit message.

Thank you for reminding, I'll read the instruction and use it next time. And sorry for the trouble.

**Pull request merged**